quantization/src/encoded_vectors_pq.rs

```diff
@@ -0,0 +1,575 @@
+            centroids = centroids.add(1);
+            lut = lut.add(centroids_count);
+        }
+        sum
+    }
+
+    #[cfg(all(target_arch = "aarch64", target_feature = "neon"))]
+    unsafe fn score_point_neon(&self, query: &EncodedQueryPQ, i: u32) -> f32 {
+        let centroids = self
+            .encoded_vectors
+            .get_vector_data(i as usize, self.metadata.vector_division.len());
+        let len = centroids.len();
+        let centroids_count = self.metadata.centroids.len();
+
+        let mut centroids = centroids.as_ptr();
+        let mut lut = query.lut.as_ptr();
+        let mut sum128 = vdupq_n_f32(0.);
+        for _ in 0..len / 4 {
+            let buffer = [
+                *lut.add(*centroids as usize),
+                *lut.add(centroids_count + *centroids.add(1) as usize),
+                *lut.add(2 * centroids_count + *centroids.add(2) as usize),
+                *lut.add(3 * centroids_count + *centroids.add(3) as usize),
+            ];
+            let c = vld1q_f32(buffer.as_ptr());
+            sum128 = vaddq_f32(sum128, c);
+
+            centroids = centroids.add(4);
+            lut = lut.add(4 * centroids_count);
+        }
+        let mut sum = vaddvq_f32(sum128);
+
+        for _ in 0..len % 4 {
+            sum += *lut.add(*centroids as usize);
+            centroids = centroids.add(1);
+            lut = lut.add(centroids_count);
+        }
+        sum
+    }
+
+    fn score_point_simple(&self, query: &EncodedQueryPQ, i: u32) -> f32 {
+        unsafe {
+            let centroids = self
+                .encoded_vectors
+                .get_vector_data(i as usize, self.metadata.vector_division.len());
+            let len = centroids.len();
+            let centroids_count = self.metadata.centroids.len();
+
+            let mut centroids = centroids.as_ptr();
+            let mut lut = query.lut.as_ptr();
+
+            let mut sum = 0.0;
+            for _ in 0..len {
+                sum += *lut.add(*centroids as usize);
+                centroids = centroids.add(1);
+                lut = lut.add(centroids_count);
+            }
+            sum
+        }
+    }
+}
+
+impl<TStorage: EncodedStorage> EncodedVectors<EncodedQueryPQ> for EncodedVectorsPQ<TStorage> {
+    fn save(&self, data_path: &Path, meta_path: &Path) -> std::io::Result<()> {
+        let metadata_bytes = serde_json::to_vec(&self.metadata)?;
+        meta_path.parent().map(std::fs::create_dir_all);
+        let mut buffer = File::create(meta_path)?;
+        buffer.write_all(&metadata_bytes)?;
+
+        data_path.parent().map(std::fs::create_dir_all);
+        self.encoded_vectors.save_to_file(data_path)?;
+        Ok(())
+    }
+
+    fn load(
+        data_path: &Path,
+        meta_path: &Path,
+        vector_parameters: &VectorParameters,
+    ) -> std::io::Result<Self> {
+        let mut contents = String::new();
+        let mut file = File::open(meta_path)?;
+        file.read_to_string(&mut contents)?;
```

Comment:
Rust has some nice [fs](https://doc.rust-lang.org/std/fs/index.html#functions) helpers that do the same. [`fs::read_to_string`](https://doc.rust-lang.org/std/fs/fn.read_to_string.html) allocates the string buffer in a smart way as well.

```suggestion
        let contents = std::fs::read_to_string(meta_path)?;
```

Reply:
Did it for the scalar quantization part too. Thanks!